scrapy/pqueues.py

```diff
@@ -0,0 +1,217 @@
+
+    def close(self):
+        active = {slot: queue.close()
+                  for slot, queue in self.pqueues.items()}
+        self.pqueues.clear()
+        return active
+
+    def __len__(self):
+        return sum(len(x) for x in self.pqueues.values()) if self.pqueues else 0
+
+    def __contains__(self, slot):
+        return slot in self.pqueues
+
+
+class DownloaderAwarePriorityQueue(object):
+
+    _DOWNLOADER_AWARE_PQ_ID = 'DOWNLOADER_AWARE_PQ_ID'
+
+    @classmethod
+    def from_crawler(cls, crawler, qfactory, startprios=None):
+        return cls(crawler, qfactory, startprios)
+
+    def __init__(self, crawler, qfactory, startprios=None):
+        ip_concurrency_key = 'CONCURRENT_REQUESTS_PER_IP'
+        ip_concurrency = crawler.settings.getint(ip_concurrency_key, 0)
+
+        if ip_concurrency > 0:
+            raise ValueError('"%s" does not support %s=%d' % (self.__class__,
+                                                              ip_concurrency_key,
+                                                              ip_concurrency))
+
+        def pqfactory(startprios=()):
+            return PriorityAsTupleQueue(qfactory, startprios)
+
+        if startprios and not isinstance(startprios, dict):
+            raise ValueError("DownloaderAwarePriorityQueue accepts "
+                             "``startprios`` as a dict; %r instance is passed."
+                             " Only a crawl started with the same priority "
+                             "queue class can be resumed." % startprios.__class__)
+        self._slot_pqueues = SlotPriorityQueues(pqfactory,
+                                                slot_startprios=startprios)
+
+        self._active_downloads = {slot: 0 for slot in self._slot_pqueues.pqueues}
+        crawler.signals.connect(self.on_response_download,
+                                signal=response_downloaded)
+        crawler.signals.connect(self.on_request_reached_downloader,
+                                signal=request_reached_downloader)
+
+    def mark(self, request):
+        meta = _get_request_meta(request)
+        if not isinstance(meta, dict):
+            raise ValueError('No meta attribute in %s' % (request, ))
+        meta[self._DOWNLOADER_AWARE_PQ_ID] = id(self)
+
+    def check_mark(self, request):
+        return request.meta.get(self._DOWNLOADER_AWARE_PQ_ID, None) == id(self)
+
+    def pop(self):
+        slots = [(active_downloads, slot)
+                 for slot, active_downloads in self._active_downloads.items()
+                 if slot in self._slot_pqueues]
+
+        if not slots:
+            return
+
+        slot = min(slots)[1]
+        request = self._slot_pqueues.pop_slot(slot)
+        self.mark(request)
+        return request
+
+    def push(self, request, priority):
+        slot = _set_scheduler_slot(request)
+        priority_slot = PrioritySlot(priority=priority, slot=slot)
+        self._slot_pqueues.push_slot(slot, request, priority_slot)
```

Comment:
can you explain why you need this  `PrioritySlot` structure, if all requests in `self._slot_pqueues.push_slot[slot]` would always have the same `priority_slot.slot`?